File: modules/common/ipc/tests/host_raw_pub_sub_test.cpp
//=================================================================================================
// Copyright (C) 2025 GRAPE Contributors
//=================================================================================================

#include <algorithm>
#include <condition_variable>
#include <random>
#include <thread>
#include <vector>

#include "catch2/catch_test_macros.hpp"
#include "grape/ipc/raw_publisher.h"
#include "grape/ipc/raw_subscriber.h"
#include "grape/ipc/session.h"

namespace {

// NOLINTBEGIN(cert-err58-cpp)

//=================================================================================================
TEST_CASE("Basic pub-sub on large message works in host-only scope", "[ipc]") {
  grape::ipc::init(grape::ipc::Config{});
  const auto* const topic = "pub_sub_test";

  // Create a large payload (eg: 1080p RGB image)
  constexpr auto PAYLOAD_SIZE = 1920U * 1080U * 3;
  auto payload = std::vector<std::byte>(PAYLOAD_SIZE);

  // Set random pattern in the payload
  std::random_device rd;
  std::mt19937 gen(rd());
  std::uniform_int_distribution<> dis(CHAR_MIN, CHAR_MAX);
  std::ranges::generate(payload,
                        [&gen, &dis]() -> std::byte { return static_cast<std::byte>(dis(gen)); });

  // define subscriber callback
  std::condition_variable recv_cond;
  std::mutex recv_mut;
  auto received_msg = std::vector<std::byte>{};
  auto pub_id = 0UL;
  const auto recv_callback = [&recv_mut, &recv_cond, &received_msg,
                              &pub_id](const grape::ipc::Sample& sample) -> void {
    const auto lk = std::lock_guard(recv_mut);
    received_msg = std::vector<std::byte>(sample.data.begin(), sample.data.end());
    pub_id = sample.info.publisher.id;
    recv_cond.notify_all();
  };

  // create pub/sub
  auto matched_sub_id = 0UL;
  const auto pub_match_cb = [&matched_sub_id](const grape::ipc::Match& match) -> void {
    matched_sub_id = match.remote_entity.id;
  };
  auto publisher = grape::ipc::RawPublisher(topic, pub_match_cb);

  auto matched_pub_id = 0UL;
  const auto sub_match_cb = [&matched_pub_id](const grape::ipc::Match& match) -> void {
    matched_pub_id = match.remote_entity.id;
  };
  auto subscriber = grape::ipc::RawSubscriber(topic, recv_callback, sub_match_cb);

  // Wait for pub/sub registration
  constexpr auto RETRY_COUNT = 10U;
  auto count_down = RETRY_COUNT;
  while ((subscriber.publisherCount() == 0) && (count_down > 0)) {
    constexpr auto REG_WAIT_TIME = std::chrono::milliseconds(200);
    std::this_thread::sleep_for(REG_WAIT_TIME);
    count_down--;
  }
  REQUIRE(subscriber.publisherCount() == 1);
  REQUIRE(matched_sub_id != 0UL);
  REQUIRE(matched_pub_id != 0UL);
  REQUIRE(matched_sub_id == subscriber.id());
  REQUIRE(matched_pub_id == publisher.id());

  // publish payload
  publisher.publish({ payload.data(), payload.size() });

  // wait a reasonable time for subscriber to receive message
  {
    constexpr auto RECV_WAIT_TIME = std::chrono::milliseconds(1000);
    auto lk = std::unique_lock(recv_mut);
    recv_cond.wait_for(lk, RECV_WAIT_TIME,
                       [&received_msg] -> bool { return not received_msg.empty(); });
  }

  // verify message
  REQUIRE(received_msg.size() == PAYLOAD_SIZE);
  REQUIRE(received_msg == payload);
  REQUIRE(pub_id == publisher.id());
}

// NOLINTEND(cert-err58-cpp)

}  // namespace

File: modules/common/ipc/tests/host_raw_pub_sub_test.py
#!/usr/bin/env python3
#=================================================================================================
# Copyright (C) 2025 GRAPE Contributors
#=================================================================================================

"""
Python equivalent of host_raw_pub_sub_test.cpp using the grape_ipc_py bindings.
Tests basic pub-sub functionality with large messages in host-only scope.
"""

import time
import random
import threading
import sys
from typing import List, Optional

try:
    import grape_ipc_py as ipc
except ImportError:
    print("ERROR: Failed to import grape_ipc_py module. Make sure it's installed or in your PYTHONPATH.")
    sys.exit(1)

def test_basic_pub_sub_large_message():
    """Test that basic pub-sub on large message works in host-only scope."""
    
    # Initialize the IPC session
    ipc.init(ipc.Config())
    
    # Topic name
    topic = "pub_sub_py_test"
    
    # Create a large payload (e.g., 1080p RGB image)
    PAYLOAD_SIZE = 1920 * 1080 * 3
    # Generate random bytes directly
    payload = bytes(random.randint(0, 255) for _ in range(PAYLOAD_SIZE))
    
    # Create synchronization event for test
    data_received = threading.Event()
    
    # Storage for received data
    received_msg = bytes()
    pub_id = 0
    
    # Define subscriber callback
    def recv_callback(sample):
        nonlocal received_msg, pub_id
        received_msg = sample.data()  
        pub_id = sample.info.publisher.id
        data_received.set()
    
    qos = ipc.QoS.BestEffort
    
    # Define match callbacks
    matched_sub_id = 0
    def pub_match_cb(match):
        nonlocal matched_sub_id
        matched_sub_id = match.remote_entity.id
    
    matched_pub_id = 0
    def sub_match_cb(match):
        nonlocal matched_pub_id
        matched_pub_id = match.remote_entity.id
    
    # Create publisher and subscriber
    publisher = ipc.RawPublisher(topic, pub_match_cb)
    subscriber = ipc.RawSubscriber(topic, qos, recv_callback, sub_match_cb)
    
    # Wait for pub/sub registration
    RETRY_COUNT = 10
    count_down = RETRY_COUNT
    while subscriber.get_publisher_count() == 0 and count_down > 0:
        time.sleep(0.2)  # 200 milliseconds
        count_down -= 1
    
    # Assert matched publishers/subscribers
    assert subscriber.get_publisher_count() == 1, "Expected 1 publisher to be matched"
    assert matched_sub_id != 0, "Expected non-zero subscriber ID"
    assert matched_pub_id != 0, "Expected non-zero publisher ID"
    
    # Publish payload
    pub_result = publisher.publish(bytes(payload))
    assert pub_result, "Failed to publish message"
    
    # Wait a reasonable time for subscriber to receive message
    success = data_received.wait(1.0)  # Wait up to 1 second
    assert success, "Timeout waiting for data to be received"
    
    # Verify message
    assert len(received_msg) == PAYLOAD_SIZE, f"Expected {PAYLOAD_SIZE} bytes, got {len(received_msg)}"
    assert received_msg == payload, "Received payload does not match sent payload"
    assert pub_id != 0, "Expected non-zero publisher ID in received sample"
    
def main():
    """Entry point for the test script."""
    try:
        test_basic_pub_sub_large_message()
        return 0
    except AssertionError as e:
        print(f"FAILURE: Assertion failed: {e}")
        return 1
    except Exception as e:
        print(f"FAILURE: Unexpected error: {e}")
        return 1

if __name__ == "__main__":
    sys.exit(main())